File: Semestrul 4/MPP/Proiecte/java_networking/Networking/src/main/java/networking/rpcprotocol/ClientRpcReflectionWorker.java
package networking.rpcprotocol;

import model.*;
import services.*;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.net.Socket;
import java.util.List;

public class ClientRpcReflectionWorker implements Runnable, IProjectObserver{
    private IProjectServices server;
    private Socket connection;

    private ObjectInputStream input;
    private ObjectOutputStream output;
    private volatile boolean connected;
    public ClientRpcReflectionWorker(IProjectServices server, Socket connection) {
        this.server = server;
        this.connection = connection;
        try {
            output = new ObjectOutputStream(connection.getOutputStream());
            output.flush();
            input = new ObjectInputStream(connection.getInputStream());
            connected = true;
        }catch (IOException e){
            e.printStackTrace();
        }
    }

    private Response handleRequest(Request request){
        Response response = null;
        String handlerName = "handle"+(request).type();
        System.out.println("HandlerName " + handlerName);
        try {
            Method method = this.getClass().getDeclaredMethod(handlerName, Request.class);
            response = (Response)method.invoke(this, request);
            System.out.println("Method " + handlerName + " invoked");
        } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException e) {
            e.printStackTrace();
        }
        return response;
    }

    private void sendResponse(Response response) throws IOException {
        System.out.println("Sending response " + response);
        synchronized (output) {
            output.writeObject(response);
            output.flush();
        }
    }

    public void run() {
        try {
            while (connected) {
                Object request = input.readObject();
                Response response = handleRequest((Request) request);
                if (response != null) {
                    sendResponse(response);
                }
            }
        } catch (IOException | ClassNotFoundException e) {
            System.out.println("Conexiune închisă" + e.getMessage());
        } finally {
            closeResources();
        }
    }

    private void closeResources() {
        try {
            if (input != null) input.close();
            if (output != null) output.close();
            if (connection != null) connection.close();
        } catch (IOException e) {
            System.out.println("Eroare la închidere resurse" + e.getMessage());
        }
    }

    @Override
    public void adauga(Donatie donatie) throws Exception {
        Response resp = new Response.Builder()
                .type(ResponseType.DONATIE_NOUA)
                .data(donatie)
                .build();
        System.out.println("Donatie adaugata: " + resp);
        try {
            sendResponse(resp);
        }catch (IOException e){
            e.printStackTrace();
        }
    }

    @Override
    public void logoutNotification() {
//        try {
//            System.out.println("Utilizatorul a fost deconectat! Te rugăm să te loghezi din nou.");
//            this.frame.dispose();
//
//            LoginWindow loginWindow = new LoginWindow();
//            loginWindow.setVisible(true);
//        } catch (IOException e) {
//            e.printStackTrace();
//        }
    }

    private static Response okResponse = new Response.Builder().type(ResponseType.OK).build();

    private Response handleLOGIN(Request request){
        System.out.println("Login request..." + request.type());
        Voluntar vDTO = (Voluntar) request.data();
        try {
            Voluntar voluntar = server.login(vDTO.getUsername(), vDTO.getPassword(), this);
            return new Response.Builder().type(ResponseType.OK).data(voluntar).build();
        }catch (Exception e){
            System.out.println("Login failed: " + e.getMessage());
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    private Response handleFIND_ALL_CAZ(Request request){
        System.out.println("Find all caz request...");
        try {
            List<Caz> cazuri = server.findAllCaz();
            return new Response.Builder().type(ResponseType.FIND_ALL_CAZ).data(cazuri).build();
        }catch (Exception e){
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    private Response handleGET_SUMA_DONATII_PT_CAZ(Request request){
        System.out.println("Get SUMA Donati PT request...");
        Integer id = (Integer) request.data();
        try {
            Integer suma = server.getSumaDonatiiPentruCaz(id);
            return new Response.Builder().type(ResponseType.GET_SUMA_DONATII_PT_CAZ).data(suma).build();
        }catch (Exception e){
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    // În ClientRpcReflectionWorker
    // În ClientRpcReflectionWorker.java
    private Response handleLIST_DTO_CAZ(Request request) {
        System.out.println("List DTO Caz request...");
        try {
            List<CazDTO> dtoCazuri = server.createCazDTOList();
            // Logging pentru depanare
            System.out.println("Serverul a generat " + dtoCazuri.size() + "cazuri DTO");
            return new Response.Builder()
                    .type(ResponseType.LIST_DTO_CAZ)
                    .data(dtoCazuri)
                    .build();
        } catch (Exception e) {
            System.err.println("Eroare la handleLIST_DTO_CAZ: " + e.getMessage());
            return new Response.Builder()
                    .type(ResponseType.ERROR)
                    .data(e.getMessage())
                    .build();
        }
    }

    private Response handleSEARCH_DONO_BYPNAME(Request request){
        System.out.println("Search Donatori by PNAME request...");
        String partialName = (String) request.data();
        try {
            List<Donator> donatori = server.searchByName(partialName);
            return new Response.Builder().type(ResponseType.SEARCH_DONO_BYPNAME).data(donatori).build();
        }catch (Exception e){
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    private Response handleSAVE_DONO(Request request){
        System.out.println("Save Donator request...");
        Donator donator = (Donator) request.data();
        try {
            Donator d1 = server.saveDonator(donator);
            return new Response.Builder().type(ResponseType.SAVE_DONO).data(d1).build();
        }catch (Exception e){
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    private Response handleFIND_DONO(Request request){
        System.out.println("Find Donator request...");
        String fullName = (String) request.data();
        try {
            Donator donator = server.findByFullName(fullName);
            return new Response.Builder().type(ResponseType.FIND_DONO).data(donator).build();
        }catch (Exception e){
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    // În ClientRpcReflectionWorker.java
    private Response handleDONATIE_NOUA(Request request) {
        System.out.println("Donatie NOUA request...");
        try {
            Donatie donatie = (Donatie) request.data();
            Donatie savedDonatie = server.saveDonatie(donatie);
            // Logging pentru confirmare
            System.out.println("Donatie salvată cu ID: " + savedDonatie.getId());
            return new Response.Builder()
                    .type(ResponseType.DONATIE_NOUA)
                    .data(savedDonatie)
                    .build();
        } catch (Exception e) {
            System.err.println("Eroare la handleDONATIE_NOUA: " + e.getMessage());
            return new Response.Builder()
                    .type(ResponseType.ERROR)
                    .data(e.getMessage())
                    .build();
        }
    }

    private Response handleDONATIE_SAVE(Request request) {
        System.out.println("Donatie SAVE request...");
        try {
            Donatie donatie = (Donatie) request.data();
            Donatie savedDonatie = server.saveDonatie(donatie);
            System.out.println("Donatie salvată cu ID: " + savedDonatie.getId());
            // Trimitem un răspuns sincronic cu un tip diferit
            return new Response.Builder()
                    .type(ResponseType.DONATIE_SAVE)
                    .data(savedDonatie)
                    .build();
        } catch (Exception e) {
            System.err.println("Eroare la handleDONATIE_SAVE: " + e.getMessage());
            return new Response.Builder()
                    .type(ResponseType.ERROR)
                    .data(e.getMessage())
                    .build();
        }
    }

    private Response handleLOGOUT(Request request){
        System.out.println("Logout request...");
        String username = (String) request.data();
        try {
            server.logout(username);
            connected = false;
            return new Response.Builder().type(ResponseType.OK).build();
        } catch (Exception e){
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }



}

File: Semestrul 4/MPP/Proiecte/c#_networking/RpcProtocol/ClientRpcReflectionWorker.cs
using System.Net.Sockets;
using System.Text;
using System.Text.Json;
using modelnet;
using networkingnet.Utils;
using servicesnet;

namespace networkingnet.RpcProtocol;

public class ClientRpcReflectionWorker : IProjectObserver
{
    private IProjectServices server;
    private TcpClient connection;

    private NetworkStream stream;
    private volatile bool connected;

    public ClientRpcReflectionWorker(IProjectServices server, TcpClient connection)
    {
        this.server = server;
        this.connection = connection;
        try
        {
            stream = connection.GetStream();
            connected = true;
        }
        catch (Exception e)
        {
            Console.WriteLine(e.StackTrace);
        }
    }
    
    public virtual void Run()
    {
        using StreamReader reader = new StreamReader(connection.GetStream());
        while (connected)
        {
            try
            {
                string requestJson = reader.ReadLine(); // Read JSON line-by-line
                if (string.IsNullOrEmpty(requestJson)) continue;
                var options = new JsonSerializerOptions
                {
                    IncludeFields = true
                };
                Console.WriteLine("Received json request {0}", requestJson);
                Request request = JsonSerializer.Deserialize<Request>(requestJson, options);
                Console.WriteLine("Deserializaed Request {0} ", request);
                Response response = handleRequest(request);
                if (response != null)
                {
                    sendResponse(response);
                }
            }
            catch (Exception e)
            {
                Console.WriteLine("run eroare {0}", e.Message);
                if (e.InnerException != null) ;
                //log.ErrorFormat("run inner error {0}", e.InnerException.Message);
                Console.WriteLine(e.StackTrace);
            }

            try
            {
                Thread.Sleep(1000);
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }
        try
        {
            stream.Close();
            connection.Close();
        }
        catch (Exception e)
        {
            Console.WriteLine(e.StackTrace);
        }
    }
    
    private static Response okResponse = JsonProtocolUtils.createOkResponse();

    private Response handleRequest(Request request)
    {
        Response response = null;
        switch (request.Type)
        {
            case RequestType.LOGIN:
                response = HandleLOGIN(request);
                break;
            case RequestType.LOGOUT:
                response = HandleLOGOUT(request);
                break;
            case RequestType.FIND_DONO:
                response = HandleFIND_DONO(request);
                break;
            case RequestType.SAVE_DONO:
                response = HandleSAVE_DONO(request);
                break;
            case RequestType.FIND_ALL_CAZ:
                response = HandleFIND_ALL_CAZ(request);
                break;
            case RequestType.LIST_DTO_CAZ:
                response = HandleLIST_DTO_CAZ(request);
                break;
            case RequestType.DONATIE_NOUA:
                response = HandleDONATIE_NOUA(request);
                break;
            case RequestType.DONATIE_SAVE:
                response = HandleDONATIE_SAVE(request);
                break;
            case RequestType.SEARCH_DONO_BYPNAME:
                response = HandleSEARCH_DONO_BYPNAME(request);
                break;
            case RequestType.GET_SUMA_DONATII_PT_CAZ:
                response = HandleGET_SUMA_DONATII_PT_CAZ(request);
                break;
            case RequestType.FIND_ALL_VOLUNTARI:
                response = HandleFIND_ALL_VOLUNTARI(request);
                break;
            default:
                Console.WriteLine("Unknown request type {0}", request.Type.ToString());
                break;
        }
        return response;
    }

    public void notifyObservers(Donatie donatie)
    {
        Console.WriteLine("notifyObservers");
        try
        {
            sendResponse(JsonProtocolUtils.createSaveDonatieResponse(donatie));
        }
        catch (IOException e)
        {
            Console.WriteLine(e.Message);
        }
    }

    private void sendResponse(Response response)
    {
        string json = JsonSerializer.Serialize(response);
        Console.WriteLine(json);
        lock (stream)
        {
            byte[] data = Encoding.UTF8.GetBytes(json + "\n");
            stream.Write(data, 0, data.Length);
            stream.Flush();
        }
    }

    private Response HandleLOGIN(Request request)
    {
        string username = request.nume;
        string password = request.password;
        try
        {
            Voluntar voluntar;
            lock (server)
                voluntar = server.login(username, password, this);
            return new Response { Type = ResponseType.OK, Voluntar = voluntar };
        }
        catch (ProjectException e)
        {
            return JsonProtocolUtils.createErrorResponse(e.Message);
        }
    }

    private Response HandleLOGOUT(Request request)
    {
        string username = request.nume;
        try
        {
            lock (server)
                server.logout(username, this);
            connected = false;
            return okResponse;
        }
        catch (ServerException e)
        {
            return JsonProtocolUtils.createErrorResponse(e.Message);
        }
    }

    private Response HandleGET_SUMA_DONATII_PT_CAZ(Request request)
    {
        int cazId = request.idCaz;
        try
        {
            int suma;
            lock (server)
                suma = server.getSumaDonatiiPentruCaz(cazId);
            return JsonProtocolUtils.createGetSumaDonatiiPtCazResponse(suma);
        }
        catch (ProjectException e)
        {
            return JsonProtocolUtils.createErrorResponse(e.Message);
        }
    }

    private Response HandleFIND_ALL_CAZ(Request request)
    {
        try
        {
            IEnumerable<Caz> cazuri;
            lock (server)
                cazuri = server.findAllCaz();
            return JsonProtocolUtils.createFindAllCazResponse(cazuri);
        }catch (ProjectException e)
        {
            return JsonProtocolUtils.createErrorResponse(e.Message);
        }
    }

    private Response HandleLIST_DTO_CAZ(Request request)
    {
        List<CazDTO> cazuri;
        try
        {
            lock (server)
                cazuri = server.createCazDTOList();
            return JsonProtocolUtils.createListDTOCazResponse(cazuri);
        }catch (ProjectException e)
        {
            return JsonProtocolUtils.createErrorResponse(e.Message);
        }
    }

    private Response HandleDONATIE_NOUA(Request request)
    {
        Donatie donatie = request.Donatie;
        try
        {
            lock (server)
                server.saveDonatie(donatie);
            return okResponse;
        }catch (ProjectException e)
        {
            return JsonProtocolUtils.createErrorResponse(e.Message);
        }
    }

    private Response HandleDONATIE_SAVE(Request request)
    {
        return null;
    }

    private Response HandleFIND_DONO(Request request)
    {
        string nume = request.fname;
        try
        {
            Donator donator;
            lock (server)
                donator = server.findByFullName(nume);
            return JsonProtocolUtils.createFindDonatorResponse(donator);
        }catch (ProjectException e)
        {
            return JsonProtocolUtils.createErrorResponse(e.Message);
        }
    }

    private Response HandleSEARCH_DONO_BYPNAME(Request request)
    {
        string nume = request.pname;
        try
        {
            IEnumerable<Donator> donatori;
            lock (server)
                donatori = server.searchByName(nume);
            return JsonProtocolUtils.createSearchDonoByPNameResponse(donatori);
        }catch (ProjectException e)
        {
            return JsonProtocolUtils.createErrorResponse(e.Message);
        }
    }

    private Response HandleSAVE_DONO(Request request)
    {
        Donator donator = request.Donator;
        try
        {
            Donator saved = server.saveDonator(donator);
            return JsonProtocolUtils.createSaveDonatorResponse(saved);
            //return okResponse;
        }catch (ProjectException e)
        {
            return JsonProtocolUtils.createErrorResponse(e.Message);
        }
    }

    private Response HandleFIND_ALL_VOLUNTARI(Request request)
    {
        try
        {
            IEnumerable<Voluntar> voluntari;
            lock (server)
                voluntari = server.findAllVoluntari();
            return JsonProtocolUtils.createFindAllVoluntariResponse(voluntari);
        }catch (ProjectException e)
        {
            return JsonProtocolUtils.createErrorResponse(e.Message);
        }
    }
    
}